File: web/src/device.ts
//! What the device in front of you remembers: `localStorage`, and the settings
//! kept in it.
//!
//! Deliberately per device and never sent to the server — a phone and a laptop
//! answer the same Set from different places, and neither has any business
//! deciding how the other draws a Diff — or how the other draws a file in Code,
//! which is the three settings at the foot of this file. Push notifications are
//! per device for the same reason, but the browser is the one that remembers
//! those, so nothing about them is kept here.
//!
//! Storage is a convenience the whole way down: a browser that refuses it costs
//! the human their drafts and their settings and nothing else, so nothing on
//! this path is worth a thrown error. Every read comes back `null` and every
//! write is dropped.

/// Where the wrap setting lives. Namespaced like the drafts beside it, so
/// everything this app leaves in a browser is legible as its own.
const WRAP = "verkstead.diff-wrap";

/// Whether Diffs are drawn wrapped on this device.
///
/// Anything but the value [`setWrapping`] writes reads as off, which is also
/// what an untouched browser answers: unwrapped is the setting nobody has
/// expressed an opinion about.
export function wrapping(): boolean {
  return read(WRAP) === "on";
}

/// Remember how this device wants Diffs drawn.
export function setWrapping(on: boolean): void {
  if (on) {
    write(WRAP, "on");
  } else {
    // Removed rather than written as "off": the absence is already the default,
    // and this way turning it off leaves nothing behind.
    forget(WRAP);
  }
}

/// And where the last repo this device made was put.
///
/// Somebody making a second repository is almost certainly putting it beside the
/// first, and where they keep their code is a fact about the machine in front of
/// them rather than something to tell the server: a laptop's `~/src` and a
/// phone's nothing at all are two answers to the same question, and neither is
/// the other's to write down.
const PARENT = "verkstead.repo-parent";

/// The directory the Create repo modal starts its browse in — empty where this
/// device has not made one yet, which is what a first run always is.
///
/// Empty rather than `null` because empty is what the field is given: a path
/// field standing empty browses the server's own home, so nothing here has to
/// know what that home is.
export function repoParent(): string {
  return read(PARENT) ?? "";
}

/// Remember where a repo was just made, for the next one.
export function setRepoParent(path: string): void {
  write(PARENT, path);
}

/// What is being held under this key, if anything — and `null` when there is no
/// storage to be had, which is a browser that blocks it or one in a context
/// that has none.
export function read(key: string): string | null {
  try {
    return localStorage.getItem(key);
  } catch {
    return null;
  }
}

/// Write `body` out, replacing whatever was under the key — and say whether it
/// landed.
///
/// Nearly every caller drops the answer, storage being a convenience: a setting
/// that could not be written is one the next visit does without. What reads it
/// is a caller holding two things of unequal worth, which is Code — its layout
/// is written apart from its unsaved text so that a storage too full for the
/// text still comes back to the splits, and a text that would not fit is taken
/// away rather than left behind to be restored over a file it no longer
/// describes (see `workbench/remembering.ts`).
export function write(key: string, body: string): boolean {
  try {
    localStorage.setItem(key, body);
    return true;
  } catch {
    // Full, or refused: what was being written is gone, and the page carries on
    // regardless.
    return false;
  }
}

/// Drop whatever is under this key.
export function forget(key: string): void {
  try {
    localStorage.removeItem(key);
  } catch {
    // As above: there was nothing to lose but the setting.
  }
}

/// And the three settings Code's own menu carries, one key each.
///
/// Namespaced like the wrap above them, and named for the editors rather than
/// for the pane: what they are about is how a file is drawn, and the pane is
/// only where they are asked for.
const EDITOR_WRAP = "verkstead.editor-wrap";
const EDITOR_SIZE = "verkstead.editor-font-size";
const EDITOR_MINIMAP = "verkstead.editor-minimap";

/// How every editor Code mounts is drawn: whether lines wrap, how large the
/// text is, and whether the minimap stands down the side.
///
/// The three [ADR 0019](../../docs/adr/0019-the-code-pane.md) exposes, and no
/// more: everything else about an editor is VS Code's default, which is what
/// asking for VS Code's feel means.
export type Drawn = {
  wrap: boolean;
  /// In pixels, which is the unit Monaco takes and the one the menu says.
  size: number;
  minimap: boolean;
};

/// VS Code's own, which is what an untouched browser answers and what a browser
/// with no storage at all answers too.
///
/// Written out rather than left to Monaco by omission, because the menu has to
/// draw a tick against one of its rows before anybody has pressed anything: a
/// setting nobody can see the value of is a row that says nothing.
export const DRAWN: Drawn = { wrap: false, size: 14, minimap: true };

/// The font sizes the menu offers, in the order it draws them.
///
/// A list rather than a field, the menu being rows: a span either side of VS
/// Code's own 14, far enough out at each end to be worth the row.
export const SIZES: readonly number[] = [10, 12, 14, 16, 18, 20];

/// How this device wants its editors drawn.
///
/// Anything but what [`setDrawn`] writes reads as the default, per setting: a
/// key somebody has edited by hand, a size no longer offered, or a browser that
/// answers nothing at all are each one setting back to VS Code's and the others
/// left alone.
export function drawn(): Drawn {
  const size = Number(read(EDITOR_SIZE));

  return {
    wrap: read(EDITOR_WRAP) === "on",
    size: SIZES.includes(size) ? size : DRAWN.size,
    minimap: read(EDITOR_MINIMAP) !== "off",
  };
}

/// Remember how this device wants them drawn, and for every editor it opens
/// after these.
///
/// Each setting left at its default is removed rather than written, the way the
/// wrap above is: the absence is already the default, so a device that has put
/// everything back leaves nothing behind.
export function setDrawn(settings: Drawn): void {
  keep(EDITOR_WRAP, settings.wrap === DRAWN.wrap, settings.wrap ? "on" : "off");
  keep(EDITOR_SIZE, settings.size === DRAWN.size, String(settings.size));
  keep(
    EDITOR_MINIMAP,
    settings.minimap === DRAWN.minimap,
    settings.minimap ? "on" : "off",
  );
}

/// Write one of them out, or take it away where it is what an untouched browser
/// already answers.
function keep(key: string, standard: boolean, body: string): void {
  if (standard) {
    forget(key);
  } else {
    write(key, body);
  }
}

File: crates/render/src/device.rs
//! What a device says it is when another one asks: the answer the identity
//! endpoint on the peer listener gives (ADR-0020).
//!
//! **The one thing on that listener nobody has to be a member to read.** A
//! device that has never been heard of dials the peer port, completes a
//! handshake without showing a certificate at all, and reads this — which is
//! what makes linking possible in the first place: the human types an address,
//! and what comes back is the device they are about to link, named and
//! fingerprinted, before anything has been agreed between the two machines.
//!
//! So it carries nothing that is anybody's work. The id is what a record and a
//! URL name this device by, and the fingerprint is the certificate that same
//! handshake just handed over, spelled to be read aloud — those two are the
//! whole of an identity, and neither of them is a secret.
//!
//! **And three things read off the machine rather than kept anywhere**: the
//! name it is shown under, the word for its operating system, and every address
//! a peer could reach it on. They are read at the moment they are answered —
//! a laptop moves between the LAN and the tailnet, and DHCP moves everybody —
//! and none of them is configured: the name is the hostname, the OS is the
//! platform's own word with *Linux (WSL)* for the one case a hostname cannot
//! tell apart, and the addresses are read off the machine's interfaces and its
//! Tailscale.
//!
//! A wire type in this crate like every other, and exported to TypeScript with
//! them: the Devices section of the Remote access pane draws this device's own
//! answer, so the browser reads the same shape a peer does.

use serde::{Deserialize, Serialize};

#[cfg(feature = "typescript")]
use ts_rs::TS;

use crate::joining::PendingJoin;

/// One device, as it answers for itself.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(TS), ts(export_to = "types.ts"))]
pub struct DeviceIdentity {
    /// The Device Id: sixteen random bytes as lower-case hex, invented at this
    /// device's first start and never changing.
    ///
    /// Spelled `device` rather than `id` because that is what the startup line
    /// calls it and what a URL segment will be — an `id` on a payload that is
    /// about a device could be a Conversation's or a Repo's.
    pub device: String,

    /// And the fingerprint of the certificate the handshake that carried this
    /// answer presented: the SHA-256 of its own bytes, upper-case hex in
    /// colon-separated pairs.
    ///
    /// Answered rather than left to be worked out from the connection, because
    /// the two are checked against each other: a caller compares what it read
    /// here with what the handshake handed it, and a device that named a
    /// certificate other than the one it presented is not the device it says it
    /// is. And it is what the human compares by eye — one person reading off a
    /// phone while another reads off a screen — which is why it is spelled the
    /// way every other tool spells one.
    pub fingerprint: String,

    /// What this device is *shown* under: the hostname of the machine it is on,
    /// read at the moment this is answered.
    ///
    /// Nothing is configured and nothing is typed. A name somebody could set
    /// would be a name two devices could be given, and what a device is *named
    /// by* is the id above — which was invented precisely because neither the
    /// hostname nor the tailnet node name can be relied on to be anybody's
    /// alone.
    ///
    /// A machine that will not say what it is called reads *this machine*,
    /// which is what the one other sentence naming this box already falls back
    /// to.
    pub name: String,

    /// And the word for the operating system it is running, which is what
    /// draws the icon beside the name.
    ///
    /// **A WSL reads *Linux (WSL)*, and that is the whole reason this is
    /// here.** A Windows machine and the WSL on it share a hostname, so the
    /// name cannot tell the two apart and this is the only thing that can —
    /// which is the setup cluster mode was written for. Everywhere else it is
    /// the platform's own spelling of itself: *Linux*, *macOS*, *Windows*.
    pub os: String,

    /// And every address a peer could reach this device on, in the order one
    /// should try them: the tailnet name and its addresses first where
    /// Tailscale is up, then the LAN addresses.
    ///
    /// **All of them, read at each answer rather than configured.** A laptop
    /// moves between the LAN and the tailnet and DHCP moves everybody, so the
    /// address somebody typed to link two devices is only the first one ever
    /// known — this list is what keeps a device that moved reachable.
    ///
    /// Empty is a device on neither a tailnet nor a network, which is an answer
    /// rather than a failure: it still has an id and a fingerprint, and those
    /// are what somebody looking at this is comparing.
    pub addresses: Vec<String>,
}

/// What a device tells each of its members when it has made its certificate
/// again (ADR-0020, *The certificate is renewed before it runs out*).
///
/// **Not a viewer type.** This one crosses the peer listener rather than the
/// workbench — it is one Verkstead telling another what it is becoming, and no
/// browser ever reads it.
///
/// **The identity whole, and the incoming fingerprint beside it.** A renewal is
/// the one call in a cluster where a device has two certificates at once, and
/// the two have to be told apart by the machine reading this or the reading is
/// worthless. So [`DeviceIdentity`] keeps the meaning it has everywhere else —
/// the fingerprint in it is the certificate this call was *made* under, which
/// the receiver checks against what its own handshake handed over, exactly as it
/// would on any other exchange — and the one being changed to is a field of its
/// own. A single fingerprint field that meant something different here would be
/// a payload the receiver had to know which call it had arrived on to read.
///
/// **And the addresses ride along**, because every device advertises all of them
/// on every exchange and this is one: a laptop that moved and renewed is reached
/// at its new addresses on the next call.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct RenewedCertificate {
    /// What this device is, as it answers anybody — with the certificate it is
    /// still *presenting* named in it, which over a changeover is the outgoing
    /// one.
    pub identity: DeviceIdentity,

    /// And the fingerprint of the certificate it is changing over to: what the
    /// receiver records against the same Device Id, and what it answers to say
    /// that it holds.
    pub incoming: String,
}

/// The **Devices** list, as the Remote access pane reads it off this machine.
///
/// The other side of [`DeviceIdentity`]: that one is what this device tells a
/// *peer* over the peer listener, and this is what it tells the browser over
/// the workbench. The same device described twice rather than two descriptions
/// of it — the identity is carried whole inside this, so the row the pane draws
/// and the answer a stranger reads cannot come to disagree about a name, an OS
/// or an address.
///
/// **Read off the machine rather than out of the settings.** Nothing here is
/// configured, which is why the pane's Devices section reads this rather than
/// the settings query the rest of the page shares — the two sections beside it
/// are read the same way and for the same reason.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(TS), ts(export_to = "types.ts"))]
pub struct DevicesView {
    /// This device, which is the row marked *this device* and the one the list
    /// always holds: a Verkstead linked to nothing still has an identity.
    pub this: DeviceIdentity,

    /// And every other device in its cluster, each as it last answered for
    /// itself — with whether it is still answering beside it.
    ///
    /// The identity is the same shape as the row above, because it is the same
    /// thing said: a member is drawn with its name, the mark for its OS and the
    /// addresses a peer could reach it on, exactly as this device is. What is
    /// different is where the answer came from — this device reads its own
    /// machine as the pane is drawn, and a member was read off the far end at
    /// the last exchange.
    ///
    /// **The count on the Remote access card comes off this**, rather than
    /// being answered beside it: there is one membership, and a number that
    /// could disagree with the rows would be two answers about it. An
    /// unreachable member counts like any other — it is linked, and a count
    /// that left it out would say the cluster had shrunk.
    pub members: Vec<LinkedDevice>,

    /// And every join this device has asked for and not yet been answered on —
    /// see [`PendingJoin`].
    ///
    /// **Beside the members rather than among them**, because a pending join is
    /// not a device: nothing has been agreed, and a row that sat in the list
    /// looking like a member would be a cluster this device had joined itself
    /// to. The count on the card is the members' alone for the same reason.
    ///
    /// Read at the moment the pane asks, the way the members are: a request
    /// whose ten minutes ran out a second ago reads expired on this answer and
    /// waiting on the one before it, and both are true when they are given.
    pub pending: Vec<PendingJoin>,
}

/// One device linked to this one: what it says it is, and whether the last dial
/// to it got through.
///
/// **Two things rather than one, because only one of them is the far end's.**
/// The identity is what that machine said about itself at the last exchange;
/// whether it is answering is this device's own finding, written by a dial that
/// worked down its addresses and reached none of them. So it sits beside the
/// identity rather than inside it — a device does not tell anybody it is
/// unreachable, and the row a peer reads off [`DeviceIdentity`] is the same
/// whichever machine is asking.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(TS), ts(export_to = "types.ts"))]
pub struct LinkedDevice {
    /// The device, as it last answered for itself.
    pub identity: DeviceIdentity,

    /// And whether the last dial to it got through. False is the row drawn
    /// dimmed, reading *unreachable* — it stays on the list, with everything
    /// about it, and an Unlink on it still works.
    pub reachable: bool,
}

/// One device nobody has typed an address for, as the **Discovered** list under
/// that same section draws it (ADR-0020, *Discovery*).
///
/// **Not a [`DeviceIdentity`], and that is the difference between the two
/// lists.** An identity is what a device *answered*, over a handshake, with the
/// fingerprint of the certificate it presented in it. Nothing here has been
/// asked anything: this is drawn off an advertisement on the LAN, which says
/// where a device is and proves nothing about it — so there is no fingerprint on
/// this row, and the string two people compare by eye is on the pending row the
/// press leaves rather than on this one.
///
/// **Three kinds of device are not in this list**: a **Member**, which is in the
/// cluster already; this device, which hears its own advertisement; and a device
/// a **Join** is already pending for, whose pending row is the answer to the
/// press somebody made. Where they are left out is the server's `discovery`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(TS), ts(export_to = "types.ts"))]
pub struct DiscoveredDevice {
    /// The Device Id off the advertisement, which is what the row is keyed by
    /// and what an **Add** on it names.
    ///
    /// Keyed by it rather than by the address, because that is the one thing
    /// about a device that is nobody else's: two Verksteads on one machine
    /// answer to one hostname on one address, and a list keyed on where
    /// something was found would draw the two of them as one row.
    pub device: String,

    /// The name it is shown under: the hostname it advertised.
    pub name: String,

    /// And the word for its operating system, which draws the mark beside the
    /// name — *Linux (WSL)* being the one thing that tells a Windows machine
    /// from the WSL on it.
    pub os: String,

    /// Where it was found, the port it advertised and all, in the order to try
    /// them.
    ///
    /// **With the port on every one of them**, unlike the addresses a member
    /// advertises: what is known here is an advertisement rather than a device's
    /// own account of itself, and the port in it is the port that device's
    /// listener really bound. It is also the only thing that tells two
    /// Verksteads on one machine apart on the page, both of them answering to
    /// one hostname at one address.
    pub addresses: Vec<String>,

    /// And how this device came to hear of it.
    pub found: Vec<FoundOn>,
}

/// Where a discovered device was found.
///
/// **A list of these on the row rather than one**, because the two sources are
/// merged by Device Id and a device that is on one LAN *and* one tailnet is
/// found twice: the row says every way this device was heard of rather than
/// whichever way was heard of first.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[cfg_attr(feature = "typescript", derive(TS), ts(export_to = "types.ts"))]
pub enum FoundOn {
    /// Heard advertising `_verkstead._tcp.local` on a network this machine is
    /// on, which is what the row reads *LAN* for.
    Lan,
}
